lightningd/pay.c

```diff
@@ -1048,13 +1049,51 @@ static void json_sendpay(struct command *cmd,
 		n_hops++;
 	}
 
-	custom_route_data = route_to_custom_route_data(cmd->ld, route);
-
 	if (n_hops == 0) {
 		command_fail(cmd, JSONRPC2_INVALID_PARAMS, "Empty route");
 		return;
 	}
 
+	/* Default (realm = 0) realm / hop_data, and other route data */
+	custom_route_data = route_to_custom_route_data(cmd->ld, route);
+
+	/* Second pass through the route input, to override realm / hop_data */
+	for (i=0, t = routetok + 1; t < end; i++, t = json_next(t)) {
+		const jsmntok_t *datatok, *realmtok;
+
+		/*
+		No need to check that t->type != JSMN_OBJECT:
+		that was already done in the first pass.
+		*/
+
+		realmtok = json_get_member(buffer, t, "realm");
+		datatok = json_get_member(buffer, t, "data");
+
+		if (realmtok) {
+			unsigned int realm;
+			if (!json_tok_number(buffer, realmtok, &realm) ||
+					realm > 0xff) {
+				command_fail(cmd, JSONRPC2_INVALID_PARAMS,
+				     "Route %zu invalid realm",
+				     i);
+				return;
+			}
+			custom_route_data.hop_data[i].realm = realm;
+		}
+
+		if (datatok) {
+			if (!hex_decode(buffer + datatok->start,
+					datatok->end - datatok->start,
+					custom_route_data.hop_data[i].per_hop_data,
+					sizeof(custom_route_data.hop_data[i].per_hop_data))) {
+				command_fail(cmd, JSONRPC2_INVALID_PARAMS,
+				     "Route %zu invalid data",
+				     i);
+				return;
+			}
+		}
+	}
+
```

Comment:
If you use a union, then how does the code that reads the data know which of the two to use? If it always uses the generic one, then you implicitly depend on the way how the C compiler structures the route_hop_bitcoin struct for its serialization. I guess you *might* get away with that if you are very careful with struct alignment, endianness and so on, but do you really want to go that way?

In the rest of the code, I see all serialization being done explicitly, without depending on how the C compiler deals with structs/unions. To me, that seems like the more sensible / portable way of doing serialization.

Reply:
I'd like to add that I allow the caller to change realm number and data independently of each other, so the user might change the realm number to signal something, but still want to use the the realm 0 format, or keep realm value 0, but use a different data format. The second might not be very useful, except maybe for our own testing purposes (it allows us to inject completely invalid realm 0 data, and see how a node responds).

Reply:
I'm not suggesting using layout assumptions: the realm byte is the descriminator.  You check that.  You can call 'generic' 'unknown' if you prefer...

Reply:
Sending a realm byte of 0 while having a hop with a different format would be confusing to BOLT-conforming implementations, so I think `realm` as discriminator, is quite fine.

Reply:
@ZmnSCPxj : If the caller passes BOLT-conforming realm 0 data, it still fails. You are really eliminating possibilities from the user; even if you don't think those possibilities will ever be useful, someone else will think they are useful to him, and try to use the API that way. It will cause unexpected failures, and maybe even a security issue for the user doing this.

Reply:
If it is BOLT-conforming realm 0 data, then the `struct route_hop_bitcoin` will work and is perfectly appropriate.  Probably if realm is 0 then `sendpay` should check the route data in detail and see that it conforms, then build the `struct route_hop_bitcoin` correctly.

Reply:
That adds way too much complexity. Why do that? To be able to use the onion? Why not just do what the user asks for? The user explicitly overrides the default data formation, so apparently the user knows what he wants, and it's different from what we'd normally do.

Let the receiving node complain if it does not understand the message.

Besides: sending "wrong" realm 0 data is useful in security / bug testing. Having this as a feature adds to "swiss army knife" usability of C-Lightning.

Reply:
Ah, I understand.